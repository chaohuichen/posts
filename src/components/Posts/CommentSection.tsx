import { Input, Textarea, Button, Typography, Grid } from "@mui/joy";
import { ChangeEvent, FormEvent, useCallback, useState } from "react";

const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;

export function isValidEmail(email: string) {
  return emailRegex.test(email);
}

const url = "https://jsonplaceholder.typicode.com/comments";
export interface CommentFormData {
  name: string;
  email: string;
  comment: string;
}

const defaultCommentDataState = {
  name: "",
  email: "",
  comment: "",
};

const CommentSection = ({
  handleCommentsInput,
}: {
  handleCommentsInput: (userInputComment: CommentFormData) => void;
}) => {
  const [formData, setFormData] = useState<CommentFormData>(defaultCommentDataState);

  const [formErrorData, setFormErrorData] = useState({
    name: "",
    email: "",
    comment: "",
  });
  const [isSubmitting, setIsSubmitting] = useState(false);

  const postCommentRequest = async (payload: CommentFormData) => {
    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        throw new Error("Network response was not ok " + response.statusText);
      }

      handleCommentsInput(payload);
      setFormData(defaultCommentDataState);
      setFormErrorData(defaultCommentDataState);
    } catch (err) {
      console.error(err);
      alert("There is something wrong, please try again later");
    }
  };
  const handlePostSubmit = useCallback(
    (event: FormEvent<HTMLFormElement>) => {
      event?.preventDefault?.();
      setIsSubmitting(true);
      const errorForm = {
        name: "",
        email: "",
        comment: "",
      };
      if (!isValidEmail(formData?.email)) {
        errorForm.email = "Please enter a valid email address.";
      }
      if (formData?.name?.length <= 0) {
        errorForm.name = "The name is required.";
      }

      if (formData?.comment?.length <= 0) {
        errorForm.comment = "The comment is required.";
      }
      if (errorForm.email !== "" || errorForm.comment !== "" || errorForm?.name !== "") {
        setFormErrorData(errorForm);
        setIsSubmitting(false);
        return;
      } else {
        postCommentRequest(formData);

        setIsSubmitting(false);
      }
    },
    [setIsSubmitting, postCommentRequest, setFormErrorData, formData]
  );

  const handleDataChange = useCallback(
    (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      setFormData((prev) => {
        return { ...prev, [e.target.name]: e.target.value };
      });

      setFormErrorData((prev) => {
        return { ...prev, [e.target.name]: "" };
      });
    },
    [setFormData, setFormErrorData]
  );

  return (
    <form onSubmit={handlePostSubmit}>
      <Grid container spacing={2} sx={{ mx: 3 }} columns={{ xs: 8, sm: 8, md: 8, lg: 8 }}>
        <Grid xs={12} sm={12} md={6} lg={3}>
          <Input
            size="lg"
            placeholder="name"
            value={formData.name}
            name="name"
            onChange={handleDataChange}
            error={formErrorData?.name !== ""}
            data-testid="nameInput"
          />
          {formErrorData?.name !== "" && (
            <Typography color="danger"> {formErrorData?.name}</Typography>
          )}
        </Grid>
        <Grid xs={12} sm={12} md={6} lg={3}>
          <Input
            size="lg"
            placeholder="email"
            value={formData.email}
            name="email"
            onChange={handleDataChange}
            error={formErrorData?.email !== ""}
            data-testid="emailInput"
          />

          {formErrorData?.email !== "" && (
            <Typography color="danger"> {formErrorData?.email}</Typography>
          )}
        </Grid>
        <Grid xs={12} sm={12} md={6} lg={6}>
          <Textarea
            placeholder="Type anythingâ€¦"
            minRows={3}
            value={formData.comment}
            name="comment"
            onChange={handleDataChange}
            error={formErrorData?.comment !== ""}
            data-testid="commentInput"
          />
          {formErrorData?.comment !== "" && (
            <Typography color="danger"> {formErrorData?.comment}</Typography>
          )}
        </Grid>
        <Grid xs={12} sm={12} md={6} lg={6} sx={{ justifyContent: "flex-end" }}>
          <Button
            type="submit"
            data-testid="submitButton"
            disabled={isSubmitting}
            sx={{ width: ["100%", 150], height: 50, my: 2, alignSelf: "flex-end" }}
          >
            Post
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};

export default CommentSection;
