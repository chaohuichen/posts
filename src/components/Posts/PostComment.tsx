import { Comment } from "@/types/Post";
import { Stack, Typography } from "@mui/joy";

const PostComment = ({ singleComment }: { singleComment: Comment }) => {
  return (
    <Stack sx={{ my: 3, mx: [0, 5] }}>
      <Stack
        direction={["column", "row"]}
        alignItems={["flex:start", "center"]}
        sx={{ my: 1 }}
      >
        <Typography level="h4" sx={{ fontSize: 14 }}>
          {singleComment?.name}
        </Typography>

        <Typography
          sx={{ ml: [0, 1], fontSize: 12 }}
          color="neutral"
        >{`(${singleComment?.email})`}</Typography>
      </Stack>
      <Typography>{singleComment?.body}</Typography>
    </Stack>
  );
};

export default PostComment;
