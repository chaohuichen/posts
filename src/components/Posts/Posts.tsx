import { Post } from "@/types/Post";
import { Select, Option } from "@mui/joy";
import { useRouter } from "next/router";

export interface PostsProps {
  posts: Post[];
}

const Posts = ({ posts }: PostsProps) => {
  const router = useRouter();

  return (
    <Select
      placeholder="Select a Post"
      sx={{ width: ["70vw", "50vw"], flex: 1 }}
      slotProps={{
        listbox: { placement: "bottom-start", sx: { minWidth: [250, 400] } },
      }}
    >
      {posts?.map((singlePost) => {
        return (
          <Option
            value={singlePost?.id}
            key={singlePost?.id}
            sx={{
              whiteSpace: "normal", // Allows wrapping
              wordBreak: "break-word", // Breaks long words
            }}
            onClick={() => router.push(`/posts/${singlePost.id}`)}
          >
            {singlePost.title}
          </Option>
        );
      })}
    </Select>
  );
};

export default Posts;
