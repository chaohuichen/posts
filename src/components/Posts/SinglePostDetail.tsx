import { Post } from "@/types/Post";
import { Stack, Typography } from "@mui/joy";

const SinglePostDetail = ({ postDetail }: { postDetail: Post }) => {
  return (
    <Stack alignItems={"center"} sx={{ mb: 0 }}>
      <Typography sx={{ width: ["100%", "80%"], my: 2, textAlign: "center" }} level="h1">
        {postDetail?.title}{" "}
      </Typography>
      <Typography sx={{ width: ["100%", "50%"] }}>{postDetail?.body} </Typography>
    </Stack>
  );
};

export default SinglePostDetail;
