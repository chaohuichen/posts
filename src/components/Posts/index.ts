import CommentSection from "./CommentSection";
import PostComment from "./PostComment";
import SinglePostDetail from "./SinglePostDetail";
export { default as Posts } from "./Posts";

export { SinglePostDetail, PostComment, CommentSection };
