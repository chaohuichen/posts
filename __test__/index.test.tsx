import { describe } from "node:test";
import HomePage, { getServerSideProps } from "../src/pages/index";
import { render, screen } from "@testing-library/react";
import { Posts } from "@/components/Posts";
import "@testing-library/jest-dom";

jest.mock("next/router", () => ({
  useRouter: jest.fn(() => ({
    push: jest.fn(),
    replace: jest.fn(),
    pathname: "/",
    route: "/",
    query: {},
    asPath: "/",
  })),
}));
// Mock the Posts component
jest.mock("@/components/Posts", () => ({
  Posts: jest.fn(() => <div>Mocked Posts Component</div>),
}));

describe("home page", () => {
  it("render the page with", async () => {
    const { container } = render(<HomePage posts={[]} isGetPostError={false} />);

    expect(container).toMatchSnapshot();
  });
  const mockPosts = [
    { id: 1, userId: 1, title: "Post 1", body: "Post 1 body" },
    { id: 2, userId: 1, title: "Post 2", body: "Post 2 body" },
  ];

  it("renders without crashing", () => {
    render(<HomePage posts={mockPosts} isGetPostError={false} />);

    // Check for the main title
    const mainTitle = screen.getByRole("heading", {
      name: /adMarketplace Post Assessment/i,
    });
    expect(mainTitle).toBeInTheDocument();

    // Check if the mocked Posts component is rendered
    expect(screen.getByText("Mocked Posts Component")).toBeInTheDocument();
  });

  it("renders posts correctly", () => {
    render(<HomePage posts={mockPosts} isGetPostError={false} />);

    // Ensure the Posts component receives the correct props
    expect(Posts).toHaveBeenCalledWith({ posts: mockPosts }, {});
  });

  it("handles empty posts gracefully", () => {
    render(<HomePage posts={[]} isGetPostError={false} />);

    // Check for the title, even if there are no posts
    const mainTitle = screen.getByRole("heading", {
      name: /adMarketplace Post Assessment/i,
    });
    expect(mainTitle).toBeInTheDocument();

    // Ensure the Posts component receives an empty array
    expect(Posts).toHaveBeenCalledWith({ posts: [] }, {});
  });

  it("displays an error state when fetching posts fails", async () => {
    const result = await getServerSideProps();

    expect(result).toEqual({
      props: { posts: [], isGetPostError: true },
    });
  });
});
