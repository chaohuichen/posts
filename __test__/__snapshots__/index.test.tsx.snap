// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the page with 1`] = `
<div>
  <div
    class="page"
  >
    <main
      class="main"
    >
      <div
        class="MuiBox-root css-1tcjf0o"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-16zwpci-JoyTypography-root"
        >
          adMarketplace Post Assessment
        </h1>
      </div>
      <div>
        Mocked Posts Component
      </div>
    </main>
  </div>
</div>
`;
